src/main/kotlin/org/rust/ide/refactoring/changeSignature/RsChangeSignatureDialog.kt

```diff
@@ -0,0 +1,370 @@
+     */
+    private fun updateConfig(): String? {
+        if (myNameField != null) {
+            val functionName = myNameField.text
+            if (validateName(functionName)) {
+                config.name = functionName
+            } else return "Function name must be a valid Rust identifier"
+        }
+
+        if (myReturnTypeField != null) {
+            val returnTypeText = myReturnTypeField.text
+            val returnType = if (returnTypeText.isBlank()) {
+                TyUnit
+            } else {
+                (myReturnTypeCodeFragment as? RsTypeReferenceCodeFragment)?.typeReference?.type
+            }
+            if (returnType != null) {
+                config.returnType = returnType
+            } else {
+                return "Function return type must be a valid Rust type"
+            }
+        }
+
+        val visField = visibilityComboBox
+        if (visField != null) {
+            if (visField.hasValidVisibility) {
+                config.visibility = visField.visibility
+            } else {
+                return "Function visibility must be a valid visibility specifier"
+            }
+        }
+
+        return null
+    }
+
+    override fun calculateSignature(): String = config.signature()
+
+    /**
+     * This is unused, since visibility is handled with a custom input.
+     */
+    override fun createVisibilityControl(): ComboBoxVisibilityPanel<String> =
+        object : ComboBoxVisibilityPanel<String>("", arrayOf()) {}
+}
+
+private fun createTypeCodeFragment(context: RsElement, type: Ty): PsiCodeFragment = RsTypeReferenceCodeFragment(
+    context.project,
+    type.render(includeTypeArguments = true, includeLifetimeArguments = true, skipUnchangedDefaultTypeArguments = true),
+    context = context
+)
+
+private fun validateName(name: String): Boolean = name.isNotBlank() && isValidRustVariableIdentifier(name)
+
+private class VisibilityComboBox(project: Project, initialVis: RsVis?, onChange: () -> Unit) {
+    private val combobox: ComboBox<String> = ComboBox<String>(createVisibilityHints(initialVis), 80)
+    private val factory: RsPsiFactory = RsPsiFactory(project)
+
+    val component: JComponent = combobox
+
+    val hasValidVisibility: Boolean
+        get() = (combobox.selectedItem as String).isBlank() || visibility != null
+    val visibility: RsVis?
+        get() = factory.tryCreateVis(combobox.selectedItem as String)
+
+    init {
+        combobox.isEditable = true
+        combobox.selectedItem = initialVis?.text.orEmpty()
+        combobox.addActionListener {
+            onChange()
+        }
+    }
+}
+
+private fun createVisibilityHints(initialVis: RsVis?): Array<String> {
+    val options = mutableListOf(
+        "",
+        "pub",
+        "pub(crate)"
+    )
+    val initialText = initialVis?.text.orEmpty()
+    if (initialText !in options) {
+        options.add(0, initialText)
+    }
+
+    return options.toTypedArray()
+}
```

Comment:
We can't just use a set, because it wouldn't have a defined order. It's probably not ideal if the order of the options was different everytime the dialog was opened.

Reply:
@Kobzol This should be OK: [`setOf`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/set-of.html) uses `java.util.LinkedHashSet` so it preserves its arguments iteration order

Reply:
Oh, I didn't know that `setOf` guarantees it. Thank you :)